### PR TITLE
DynRankView: make layout() return the same as a corresponding static View

### PR DESCRIPTION
**Pull request by PhilMiller**

### CODE REVIEW

**Comment by PhilMiller** [comment]
Replaces https://github.com/kokkos/kokkos/pull/4775

File: containers/src/Kokkos_DynRankView.hpp [hunk 1]
```diff
@@ -477,10 +477,7 @@ class DynRankView : public ViewTraits<DataType, Properties...> {
     return static_cast<int>(m_map.extent(r));
   }
 
-  KOKKOS_INLINE_FUNCTION constexpr typename traits::array_layout layout()
-      const {
-    return m_map.layout();
-  }
+  KOKKOS_INLINE_FUNCTION constexpr typename traits::array_layout layout() const;
```
Comment (dalg24) [inline containers/src/Kokkos_DynRankView.hpp:480]: Why did you move the definition out of the class?

**Review by Rombur — APPROVED** [review]

**Comment by PhilMiller** [comment]
#5188 potentially unblocks this

**Comment by PhilMiller** [comment]
This test code that I needed to pass now passes:

```c++
    Kokkos::View<double****> u("foo", 1, 2, 3, 4);

    auto ulayout = u.layout();
    ASSERT_EQ(ulayout.dimension[0], 1);
    ASSERT_EQ(ulayout.dimension[1], 2);
    ASSERT_EQ(ulayout.dimension[2], 3);
    ASSERT_EQ(ulayout.dimension[3], 4);
    ASSERT_EQ(ulayout.dimension[4], KOKKOS_INVALID_INDEX);
    ASSERT_EQ(ulayout.dimension[5], KOKKOS_INVALID_INDEX);
    ASSERT_EQ(ulayout.dimension[6], KOKKOS_INVALID_INDEX);
    ASSERT_EQ(ulayout.dimension[7], KOKKOS_INVALID_INDEX);
```

File: containers/src/Kokkos_DynRankView.hpp (hunk 1 above)
Comment (PhilMiller) [reply containers/src/Kokkos_DynRankView.hpp:480]: The new definition depends on the free function `Impl::as_view_of_rank_n` defined below. I don't want to pile on a bunch of forward declarations for the sake of keeping this method definition in-class.

**Comment by PhilMiller** [comment]
retest this please

**Review by masterleinad — COMMENTED** [review]
Why is it advantageous for `DynRankView` to return the layout of the corresponding `View` instead of what it does currently?

**Comment by PhilMiller** [comment]
It allows easy construction of a View or DRV  with matching rank and extent, and thus deep_copy between them.

**Comment by PhilMiller** [comment]
Also, generally, it's weird that the rank of the DRV doesn't match the rank described by its layout

**Review by masterleinad — APPROVED** [review]
Yes, I think this makes more sense than always returning a 7-dim layout.
